File: 01-array-string/22-zigzag-conversion/Solution.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s
        rows = [''] * numRows
        cur, step = 0, 1
        for c in s:
            rows[cur] += c
            if cur == 0:
                step = 1
            elif cur == numRows - 1:
                step = -1
            cur += step
        return ''.join(rows)

File: 01-array-string/22-zigzag-conversion/Solution.java
class Solution {
    public String convert(String s, int numRows) {
        if (numRows == 1) return s;
        StringBuilder[] rows = new StringBuilder[numRows];
        for (int i = 0; i < numRows; i++) rows[i] = new StringBuilder();
        int cur = 0, step = 1;
        for (char c : s.toCharArray()) {
            rows[cur].append(c);
            if (cur == 0) step = 1;
            else if (cur == numRows - 1) step = -1;
            cur += step;
        }
        StringBuilder res = new StringBuilder();
        for (StringBuilder row : rows) res.append(row);
        return res.toString();
    }
}
